devito/finite_differences/coefficients.py

```diff
@@ -0,0 +1,191 @@
+        for d in self.data:
+            function_list += (d.function,)
+        return list(set(function_list))
+
+    def rules(self):
+
+        def generate_subs(d):
+
+            deriv_order = d.deriv_order
+            function = d.function
+            dim = d.dimension
+            coeffs = d.coefficients
+
+            fd_order = len(coeffs)-1
+
+            side = function.get_side(dim, deriv_order)
+            stagger = function.get_stagger(dim, deriv_order)
+
+            subs = {}
+
+            indices, x0 = generate_indices(function, dim, dim.spacing, fd_order,
+                                           side=side, stagger=stagger)
+
+            for j in range(len(coeffs)):
+                subs.update({function.fd_coeff_symbol()
+                             (indices[j], deriv_order, function, dim): coeffs[j]})
+
+            return subs
+
+        # Figure out when symbolic coefficients can be replaced
+        # with user provided coefficients and, if possible, generate
+        # replacement rules
+        rules = {}
+        for d in self.data:
+            if isinstance(d.coefficients, np.ndarray):
+                rules.update(generate_subs(d))
+
+        return rules
+
+
+class Coefficient(object):
+    """
+    Prepare custom FD coefficients to pass to Coefficients object.
+
+    Parameters
+    ----------
+    deriv_order : represents the order of the derivative being taken.
+    function : represents the function for which the supplied coefficients
+               will be used.
+    dimension : represents the dimension with respect to which the
+                derivative is being taken.
+    coefficients : represents the set of finite difference coefficients
+                  intended to be used in place of the standard
+                  coefficients (obtained from a Taylor expansion).
+    """
+
+    def __init__(self, deriv_order, function, dimension, coefficients):
+
+        self.check_input(deriv_order, function, dimension, coefficients)
+
+        # Ensure the given set of coefficients is the correct length
+        if dimension.is_Time:
+            if len(coefficients)-1 != function.time_order:
+                raise ValueError("Number FD weights provided does not "
+                                 "match the functions space_order")
+        elif dimension.is_Space:
+            if len(coefficients)-1 != function.space_order:
+                raise ValueError("Number FD weights provided does not "
+                                 "match the functions space_order")
+
+        self.is_Coefficient = True
+
+        self.deriv_order = deriv_order
+        self.function = function
+        self.dimension = dimension
+        self.coefficients = coefficients
+
+    def check_input(self, deriv_order, function, dimension, coefficients):
+        if not isinstance(deriv_order, int):
+            raise TypeError("Derivative order must be an integer")
+        if not function.is_Function:
```

Comment:
probably better off to use isinstance, otherwise you might get an AttribtueError

Reply:
Yeah, I see. But `isintance(function, Function)` won't work by default and I'm not sure how it should be imported to get it to work?

Reply:
Using try/except for now(?).